File: 322.coin-change.py
# https://leetcode-cn.com/problems/coin-change/
#
# algorithms
# Medium (29.48%)
# Total Accepted:    6,546
# Total Submissions: 22,206

from collections import deque


class Solution(object):
    def coinChange(self, coins, amount):
        """
        :type coins: List[int]
        :type amount: int
        :rtype: int
        """
        if amount == 0:
            return 0

        coins.sort(reverse=True)
        queue = deque()
        s = set()

        for c in coins:
            if c == amount:
                return 1
            queue.append((c, 1))
            s.add(c)

        while len(queue) > 0:
            cur_s, cur_n = queue.popleft()
            for c in coins:
                tmp = cur_s + c
                if tmp == amount:
                    return cur_n + 1
                elif tmp < amount and tmp not in s:
                    queue.append((tmp, cur_n + 1))
                    s.add(tmp)

        return -1

File: 322.Coin-Change.go
// https://leetcode.com/problems/coin-change/
//
// algorithms
// Medium (30.09%)
// Total Accepted:    188,159
// Total Submissions: 625,420

package leetcode

var MAX_INT = 1<<31 - 1

func coinChange(coins []int, amount int) int {
	dp := make([]int, amount+1)
	for i := 1; i <= amount; i++ {
		dp[i] = MAX_INT
	}

	for i := 0; i < amount; i++ {
		for _, c := range coins {
			if i+c <= amount && dp[i]+1 < dp[i+c] {
				dp[i+c] = dp[i] + 1
			}
		}
	}

	if dp[amount] == MAX_INT {
		return -1
	}

	return dp[amount]
}
